opds_import.py

```diff
@@ -173,6 +176,116 @@ def from_config(cls, _db, collection=None):
+            for i in lookup._run_collection_self_tests():
+                yield i
+
+    def _run_collection_self_tests(self):
+        """Run the self-test suite on the Collection associated with this
+        MetadataWranglerOPDSLookup.
+        """
+        if not self.collection:
+            return
+        metadata_identifier = None
+        try:
+            metadata_identifier = self.collection.metadata_identifier
+        except ValueError, e:
+            # This collection has no metadata identifier. It's
+            # probably a "Manual intervention" collection. It cannot
+            # interact with the metadata wrangler and there's no need
+            # to test it.
+            return
+
+        # Check various endpoints that yield OPDS feeds.
+        one_day_ago = datetime.datetime.utcnow() - datetime.timedelta(days=1)
+        for title, m, args in (
+            (
+                "Metadata updates in last 24 hours", 
+                self.updates, [one_day_ago]
+            ),
+            (
+                "Titles where we could (but haven't) provide information to the metadata wrangler",
+                self.metadata_needed, []
+            )
+        ):
+            yield self._feed_self_test(title, m, *args)
+
+    def _feed_self_test(self, name, method, *args):
+        """Retrieve a feed from the metadata wrangler and 
+        turn it into a SelfTestResult.
+        """
+        result = SelfTestResult(name)
+        result.collection = self.collection
+
+        # If the server returns a 500 error we don't want to raise an
+        # exception -- we want to record it as part of the test
+        # result.
+        kwargs = dict(allowed_response_codes=['%sxx' % f for f in range(1,6)])
+
+        response = method(*args, **kwargs)
+        self._annotate_feed_response(result, response)
+
+        # We're all done.
+        result.end = datetime.datetime.utcnow()
+        return result
+
+    @classmethod
+    def _annotate_feed_response(cls, result, response):
+        """Parse an OPDS feed and annotate a SelfTestResult with some
+        information about it:
+
+        * How the feed was requested.
+        * What the response code was.
+        * The number of items on the first page.
+        * The title of each item on the page, if any.
+        * The total number of items in the feed, if available.
+
+        :param result: A SelfTestResult object.
+        :param response: A requests Response object.
+        """
+        lines = []
+        lines.append("Request URL: %s" % response.url)
+        lines.append(
+            "Request authorization: %s" %
+            response.request.headers.get('Authorization')
+        )
+        lines.append("Status code: %d" % response.status_code)
+        result.success = response.status_code == 200
+        if result.success:
+            feed = feedparser.parse(response.content)
+            total_results = feed['feed'].get('opensearch_totalresults')
```

Comment:
This feature doesn't exist yet on the metadata wrangler side -- that's https://jira.nypl.org/browse/SIMPLY-2422.